src/renderer/SplatRenderer.cpp

```diff
@@ -0,0 +1,259 @@
+        viewportData.mFallbackBytes.size(), cudaMemcpyDeviceToHost));
+    glNamedBufferSubData(viewportData.mImageBuffer, 0, viewportData.mFallbackBytes.size(),
+        viewportData.mFallbackBytes.data());
+  }
+
+  // Copy image contents to frame buffer. For this, we disable the depth test, depth writing, and
+  // perform pre-multiplied alpha blending.
+  mCopyShader.Bind();
+
+  glDisable(GL_DEPTH_TEST);
+  glEnable(GL_BLEND);
+  glBlendFunc(GL_ONE, GL_ONE_MINUS_SRC_ALPHA);
+  glDepthMask(GL_FALSE);
+
+  glUniform1i(mUniforms.mWidth, viewportData.mWidth);
+  glUniform1i(mUniforms.mHeight, viewportData.mHeight);
+
+  glBindBufferBase(GL_SHADER_STORAGE_BUFFER, 0, viewportData.mImageBuffer);
+  glDrawArrays(GL_TRIANGLE_STRIP, 0, 4);
+
+  glEnable(GL_DEPTH_TEST);
+  glDisable(GL_BLEND);
+  glDepthMask(GL_TRUE);
+
+  mCopyShader.Release();
+
+  if (cudaPeekAtLastError() != cudaSuccess) {
+    logger().error(
+        "A CUDA error occurred during rendering: {}", cudaGetErrorString(cudaGetLastError()));
+  }
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////
+
+SplatRenderer::ViewportData& SplatRenderer::getCurrentViewportData() {
+
+  std::array<GLint, 4> viewportExtent{};
+  glGetIntegerv(GL_VIEWPORT, viewportExtent.data());
+  GLint width  = viewportExtent.at(2);
+  GLint height = viewportExtent.at(3);
+
+  auto vistaViewport   = GetVistaSystem()->GetDisplayManager()->GetCurrentRenderInfo()->m_pViewport;
+  auto viewportData    = mViewportData.find(vistaViewport);
+  bool needsRecreation = false;
+
+  // If we haven't seen this viewport before or it changed size, we need to create new resources for
+  // it.
+  if (viewportData == mViewportData.end()) {
+    needsRecreation = true;
+
+  } else if (viewportData->second.mWidth != width || viewportData->second.mHeight != height) {
+
+    // If the viewport changed size, we have to free any previously created resources first.
+    if (!viewportData->second.mInteropFailed) {
+      cudaGraphicsUnregisterResource(viewportData->second.mImageBufferCuda);
+    } else {
+      cudaFree(viewportData->second.mFallbackBufferCuda);
+    }
+    glDeleteBuffers(1, &viewportData->second.mImageBuffer);
+
+    needsRecreation = true;
+  }
+
+  // Create OpenGL buffer ready for CUDA/GL interop.
+  if (needsRecreation) {
+
+    ViewportData data;
+
+    bool useInterop = true;
+
+    glCreateBuffers(1, &data.mImageBuffer);
+    glNamedBufferStorage(
+        data.mImageBuffer, width * height * 4 * sizeof(float), nullptr, GL_DYNAMIC_STORAGE_BIT);
+
+    if (useInterop) {
+      if (cudaPeekAtLastError() != cudaSuccess) {
+        logger().error(
+            "A CUDA error occurred in setup: {}", cudaGetErrorString(cudaGetLastError()));
+      }
+      cudaGraphicsGLRegisterBuffer(
+          &data.mImageBufferCuda, data.mImageBuffer, cudaGraphicsRegisterFlagsWriteDiscard);
+      useInterop &= (cudaGetLastError() == cudaSuccess);
+    }
```

Comment:
`useInterop` is always true here.

```suggestion
    if (cudaPeekAtLastError() != cudaSuccess) {
      logger().error(
          "A CUDA error occurred in setup: {}", cudaGetErrorString(cudaGetLastError()));
    }
    cudaGraphicsGLRegisterBuffer(
        &data.mImageBufferCuda, data.mImageBuffer, cudaGraphicsRegisterFlagsWriteDiscard);
    useInterop &= (cudaGetLastError() == cudaSuccess);
```